integration/client/container_idmapped_linux_test.go

```diff
@@ -0,0 +1,121 @@
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package client
+
+import (
+	"os"
+	"strings"
+	"syscall"
+	"testing"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/oci"
+	"github.com/containerd/containerd/snapshots/overlay/overlayutils"
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+func TestIDMappedOverlay(t *testing.T) {
+	var (
+		upperPath   string
+		lowerPaths  []string
+		snapshotter = "overlayfs"
+		ctx, cancel = testContext(t)
+		id          = t.Name()
+	)
+	defer cancel()
+
+	if ok, err := overlayutils.SupportsIDMappedMounts(); err != nil || !ok {
+		t.Skip("overlayfs doesn't support idmapped mounts")
+	}
+
+	client, err := newClient(t, address)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+
+	image, err := client.Pull(ctx, testMultiLayeredImage, containerd.WithPullUnpack)
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Logf("image %s pulled!", testMultiLayeredImage)
+
+	hostID := uint32(33)
+	contID := uint32(0)
+	length := uint32(65536)
+
+	uidMap := specs.LinuxIDMapping{
+		ContainerID: contID,
+		HostID:      hostID,
+		Size:        length,
+	}
+	gidMap := specs.LinuxIDMapping{
+		ContainerID: contID,
+		HostID:      hostID,
+		Size:        length,
+	}
+
+	container, err := client.NewContainer(ctx, id,
+		containerd.WithImage(image),
+		containerd.WithImageConfigLabels(image),
+		containerd.WithSnapshotter(snapshotter),
+		containerd.WithNewSnapshot(id, image, containerd.WithRemapperLabels(uidMap.ContainerID, uidMap.HostID, gidMap.ContainerID, gidMap.HostID, length)),
+		containerd.WithNewSpec(oci.WithImageConfig(image),
+			oci.WithUserNamespace([]specs.LinuxIDMapping{uidMap}, []specs.LinuxIDMapping{gidMap}),
+			longCommand))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer container.Delete(ctx, containerd.WithSnapshotCleanup)
+
+	t.Logf("container %s created!", id)
+	o := client.SnapshotService(snapshotter)
+	mounts, err := o.Mounts(ctx, id)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	m := mounts[0]
+	if m.Type != "overlay" {
+		t.Fatalf("invalid mount -- %s; expected %s", m.Type, snapshotter)
+	}
+
+	for _, o := range m.Options {
+		if strings.HasPrefix(o, "upperdir=") {
+			upperPath = strings.TrimPrefix(o, "upperdir=")
+		} else if strings.HasPrefix(o, "lowerdir=") {
+			lowerPaths = strings.Split(strings.TrimPrefix(o, "lowerdir="), ",")
+		}
+	}
+
+	t.Log("check lowerdirs")
+	for _, l := range lowerPaths {
+		if _, err := os.Stat(l); err == nil {
+			t.Fatalf("lowerdir=%s should not exist", l)
+		}
+	}
+
+	t.Logf("check stats of uppedir=%s", upperPath)
+	st, err := os.Stat(upperPath)
+	if err != nil {
+		t.Fatalf("failed to stat %s", upperPath)
+	}
+	stat := st.Sys().(*syscall.Stat_t)
```

Comment:
This will generate a panic if it is not that type. See this: https://go.dev/tour/methods/15

Can you please change it to _test_ if it is that type (as explained on the link), and if it is not just make the test fail?

I understand this should not happen in practice, but let's handle errors nicely :)

Reply:
Yeah, it makes sense, thanks!

Reply:
Done